templates/module.build.ps1

```diff
@@ -103,13 +103,22 @@ task Publish {
 task PublishNuget {
     $newVersion = New-Object version -ArgumentList 1, 0, 0, $BuildNumber
     "Version is $newVersion"
-    $Public  = @(Get-ChildItem -Path $ModulePath\Public\*.ps1 -ErrorAction SilentlyContinue)
-    $Functions = $public.basename
-    Update-ModuleManifest -Path $ModulePath\$ModuleName.psd1 -ModuleVersion $newVersion -FunctionsToExport $Functions
+
+    $PublicScriptPath = "$ModulePath\Public"
+    $psFilter = "*.ps1"
+    $functions = ( Get-ChildItem -Path $PublicScriptPath -Filter $psFilter -Recurse -ErrorAction SilentlyContinue | Select-String -Pattern "function" )
+    $exportFunctions = @()
+    foreach ($func in $functions) {
+        $x = $func.Line -replace 'function '
```

Comment:
I was checking out the possible syntax for functions [here](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions?view=powershell-6 ) and there are a number of options
1. function test {}
1. function test ($param1) {}
1. function test
{
}

So I started thinking we might run into issues depending what syntax people use. That got me back to my RegEx idea :-)

What about replacing [lines 112-114](https://github.com/tabs-not-spaces/TikabuPSModuleTemplate/blob/bf0da3ec2ec37f25f47d4dea5e80edfa1f56e3ee/templates/module.build.ps1#L112-L114) with this?

```PowerShell
if ($func.Line -match 'function\s+(.*?)(?:[\s({].*|$)') {
  $exportFunction += $matches[1]
}
```
I know the regex is not nice but it seems to match all the conditions.

Reply:
I had an idea about a better way to do this. We can use the AST and scrap the regex. It will ensure we pick up all functions.

```PowerShell
$files = Get-ChildItem -Path $PublicScriptPath -Filter $psFilter -Recurse -ErrorAction SilentlyContinue
$exportFunctions = @()
foreach ($file in $files) {

    $tokens = $errors = $null
    $ast = [System.Management.Automation.Language.Parser]::ParseFile(
        $file.fullname,
        [ref]$tokens,
        [ref]$errors)
    
        $functionDefinitions = $ast.FindAll({
            param([System.Management.Automation.Language.Ast] $Ast)
        
            $Ast -is [System.Management.Automation.Language.FunctionDefinitionAst] -and
            # Class methods have a FunctionDefinitionAst under them as well, but we don't want them.
            ($PSVersionTable.PSVersion.Major -lt 5 -or
            $Ast.Parent -isnot [System.Management.Automation.Language.FunctionMemberAst])
        
        }, $true)
    
        $exportFunction += $functionDefinitions.Name
}